packages/melody-hooks/src/component.js

```diff
@@ -293,52 +293,71 @@ Object.assign(Component.prototype, {
         // `setState` is called inside the component function
         this.isRunningComponentFn = true;
 
+        // When a component function throws an error we store
+        // it in this variable, break the loop, clean up and
+        // finally throw the error.
+        let error;
+
         // tracks if we at least run the component function once
         let updated = false;
 
         // tracks how often we did call the component function
         // due to calls to `setState` in the component function.
         let i = 0;
 
-        while (
-            // The state has changed
-            this.flushState() ||
-            // We received new props
-            this.isPropsDirty ||
-            // The component function was never invoked
-            this.isCollectingHooks
-        ) {
-            if (i++ > RENDER_LIMIT) {
-                throw new Error(
-                    'Too many re-renders. Melody limits the number of renders to prevent ' +
-                        'an infinite loop.'
-                );
+        try {
+            while (
+                // The state has changed
+                this.flushState() ||
+                // We received new props
+                this.isPropsDirty ||
+                // The component function was never invoked
+                this.isCollectingHooks
+            ) {
+                if (i++ > RENDER_LIMIT) {
+                    throw new Error(
+                        'Too many re-renders. Melody limits the number of renders to prevent ' +
+                            'an infinite loop.'
+                    );
+                }
+                // the state is now not considered dirty anymore
+                this.isStateDirty = false;
+                // the props is now not considered dirty anymore
+                this.isPropsDirty = false;
+
+                // Reset the hooks pointer
+                this.hooksPointer = -1;
+                // Run the component functions
+
+                this.data = this.componentFn(this.props) || {};
+
+                // Mark that we have already collected the hooks
+                // This only may happen once.
+                this.isCollectingHooks = false;
```

Comment:
still not sure about this part. if an error is thrown during `this.componentFn` we will not set `this.isCollectingHooks` to false. If the component is rendered again, there might be new errors.